enzyme/Enzyme/MLIR/Implementations/MemRefAutoDiffOpInterfaceImpl.cpp

```diff
@@ -0,0 +1,106 @@
+
+#include "Implementations/CoreDialectsAutoDiffImplementations.h"
+#include "Interfaces/AutoDiffOpInterface.h"
+#include "Interfaces/AutoDiffTypeInterface.h"
+#include "Interfaces/GradientUtils.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/IR/DialectRegistry.h"
+#include "mlir/Support/LogicalResult.h"
+
+using namespace mlir;
+using namespace mlir::enzyme;
+
+namespace {
+struct LoadOpInterface
+    : public AutoDiffOpInterface::ExternalModel<LoadOpInterface,
+                                                memref::LoadOp> {
+  LogicalResult createForwardModeAdjoint(Operation *op, OpBuilder &builder,
+                                         MGradientUtils *gutils) const {
+    auto loadOp = cast<memref::LoadOp>(op);
+    if (!gutils->isConstantValue(loadOp)) {
+      SmallVector<Value> inds;
+      for (auto ind : loadOp.getIndices())
+        inds.push_back(gutils->getNewFromOriginal(ind));
+      mlir::Value res = builder.create<memref::LoadOp>(
+          loadOp.getLoc(), gutils->invertPointerM(loadOp.getMemref(), builder),
+          inds);
+      gutils->setDiffe(loadOp, res, builder);
+    }
+    gutils->eraseIfUnused(op);
+    return success();
+  }
+};
+
+struct StoreOpInterface
+    : public AutoDiffOpInterface::ExternalModel<StoreOpInterface,
+                                                memref::StoreOp> {
+  LogicalResult createForwardModeAdjoint(Operation *op, OpBuilder &builder,
+                                         MGradientUtils *gutils) const {
+    auto storeOp = cast<memref::StoreOp>(op);
+    if (!gutils->isConstantValue(storeOp.getMemref())) {
+      SmallVector<Value> inds;
+      for (auto ind : storeOp.getIndices())
+        inds.push_back(gutils->getNewFromOriginal(ind));
+      builder.create<memref::StoreOp>(
+          storeOp.getLoc(), gutils->invertPointerM(storeOp.getValue(), builder),
+          gutils->invertPointerM(storeOp.getMemref(), builder), inds);
+    }
+    gutils->eraseIfUnused(op);
+    return success();
+  }
+};
+
+struct AllocOpInterface
+    : public AutoDiffOpInterface::ExternalModel<AllocOpInterface,
+                                                memref::AllocOp> {
+  LogicalResult createForwardModeAdjoint(Operation *op, OpBuilder &builder,
+                                         MGradientUtils *gutils) const {
+    auto allocOp = cast<memref::AllocOp>(op);
+    if (!gutils->isConstantValue(allocOp)) {
+      BlockAndValueMapping map;
+      for (auto op : allocOp->getOperands())
+        map.map(op, gutils->getNewFromOriginal(op));
+      auto nop = builder.clone(*allocOp, map);
+      gutils->setDiffe(allocOp, nop->getResult(0), builder);
+    }
+    gutils->eraseIfUnused(op);
+    return success();
+  }
+};
+
+class MemRefTypeInterface
+    : public AutoDiffTypeInterface::ExternalModel<MemRefTypeInterface,
+                                                  MemRefType> {
+public:
+  Value createNullValue(Type self, OpBuilder &builder, Location loc) const {
```

Comment:
If you want a hacky solution, we can do `llvm.mlir.null` to get a null pointer and then `polygeist.pointer2memref`.